src/main/kotlin/io/github/anki/anki/service/CardsService.kt

```diff
@@ -9,56 +9,68 @@ import io.github.anki.anki.service.model.Pagination
 import io.github.anki.anki.service.model.mapper.toCard
 import io.github.anki.anki.service.model.mapper.toMongo
 import org.bson.types.ObjectId
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
 import org.springframework.stereotype.Service
+import reactor.core.publisher.Flux
+import reactor.core.publisher.Mono
 
 @Service
 class CardsService(
     private val cardRepository: CardRepository,
     private val deckService: DeckService,
 ) {
-    fun createNewCard(userId: String, card: Card): Card {
-        deckService.validateUserHasPermissions(card.deckId, userId)
-        return cardRepository
-            .insert(card.toMongo())
-            .get()
-            .toCard()
-    }
+    fun createNewCard(userId: String, card: Card): Mono<Card> =
+        deckService
+            .validateUserHasPermissions(card.deckId, userId)
+            .flatMap { cardRepository.insert(card.toMongo()) }
+            .map(MongoCard::toCard)
 
-    fun findCardsByDeckWithPagination(deckId: String, userId: String, pagination: Pagination): List<Card> {
+    fun findCardsByDeckWithPagination(deckId: String, userId: String, pagination: Pagination): Flux<Card> =
         deckService.validateUserHasPermissions(deckId, userId)
-        return cardRepository
-            .findByDeckIdWithStatus(
-                deckId = ObjectId(deckId),
-                limit = pagination.limit,
-                offset = pagination.offset,
-            )
-            .get()
-            .map { it.toCard() }
-    }
+            .flatMapMany {
+                cardRepository
+                    .findByDeckIdWithStatus(
+                        deckId = ObjectId(deckId),
+                        limit = pagination.limit,
+                        offset = pagination.offset,
+                    )
+            }
+            .map(MongoCard::toCard)
 
-    fun updateCard(userId: String, card: Card): Card {
-        deckService.validateUserHasPermissions(card.deckId, userId)
-        val mongoCard: MongoCard = getCardById(card.id ?: throw IllegalArgumentException("Card Id can not be null"))
-        val updatedMongoCard: MongoCard = mongoCard.update(card)
-        if (mongoCard == updatedMongoCard) {
-            return mongoCard.toCard()
-        }
-        return cardRepository.save(updatedMongoCard).get().toCard()
+    fun updateCard(userId: String, card: Card): Mono<Card> {
+        card.id ?: return Mono.error(IllegalArgumentException("Card Id can not be null"))
+        return deckService.validateUserHasPermissions(card.deckId, userId)
+            .flatMap { getCardById(card.id) }
+            .flatMap { saveIfNotEquals(it, card) }
```

Comment:
```kotlin
return Mono.just(card)
     .mapNotNull { it.id }
     .switchIfEmpty { IllegalArgumentException("Card Id can not be null") } 
     .flatMap { id -> getCardById(id) }
     .flatMap { saveIfNotEquals(it, card) }
```

Reply:
it is not smart casted to String, and i have to validate id is not String? anyway

Reply:
You don't need the validation because `mapNotNull` guarantee that the value is not null.
You can use `!!` operator. It'd ok here